File: 114. Flatten Binary Tree to Linked List/main.c
#include <stdio.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void flatten(struct TreeNode *root) {
    if (root == NULL) {
        return;
    }

    if (root->right != NULL) {
        flatten(root->right);
    }

    if (root->left != NULL) {
        flatten(root->left);
        struct TreeNode *node = root->left;

        while (node->right != NULL) {
            node = node->right;
        }

        node->right = root->right;
        root->right = root->left;
        root->left = NULL;
    }
}

static struct TreeNode *flat(struct TreeNode *node) {
    if (node == NULL) {
        return NULL;
    }

    struct TreeNode *right = node->right;
    node->right = flat(node->left);
    node->left = NULL;

    struct TreeNode *current = node;

    while (current->right != NULL) {
        current = current->right;
    }

    current->right = flat(right);

    return node;
}

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void flatten(struct TreeNode *root) {
    flat(root);
}

File: 114. Flatten Binary Tree to Linked List/main.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return

        if root.right:
            self.flatten(root.right)

        if root.left:
            self.flatten(root.left)

            node = root.left

            while node.right:
                node = node.right

            node.right = root.right
            root.right = root.left
            root.left = None


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flat(self, node: Optional[TreeNode]) -> Optional[TreeNode]:
        if not node:
            return None

        right = node.right
        node.right = self.flat(node.left)
        node.left = None
        current = node

        while current.right:
            current = current.right

        current.right = self.flat(right)

        return node

    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        self.flat(root)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.prev = None

    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return

        self.flatten(root.right)
        self.flatten(root.left)
        root.right = self.prev
        root.left = None
        self.prev = root
